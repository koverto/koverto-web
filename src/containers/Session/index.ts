export { SessionProvider } from "./provider"
export { useSession } from "./context"
