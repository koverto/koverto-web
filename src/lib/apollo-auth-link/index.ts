import { ApolloLink } from "apollo-link"
import { ContextSetter, setContext } from "apollo-link-context"

export class AuthLink extends ApolloLink {
  constructor(private tokenGetter: () => string) {
    super()

    this.request = setContext(this.setHeaders).request
  }

  public setHeaders: ContextSetter = (_, { headers = {} }) => {
    const token = this.tokenGetter()

    if (token) {
      headers.authorization = `Bearer ${token}`
    }

    return headers
  }
}
