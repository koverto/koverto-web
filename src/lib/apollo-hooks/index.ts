import { OperationVariables as _OperationVariables } from "@apollo/react-common"
import { DocumentNode } from "graphql"

export abstract class Operation<D = unknown, V = _OperationVariables> {
  protected data: D
  protected variables: V

  constructor(public statement: DocumentNode) {}
}

export type OperationData<
  T extends Operation<unknown, unknown>
> = T extends Operation<infer D, unknown> ? D : never

export type OperationVariables<
  T extends Operation<unknown, unknown>
> = T extends Operation<unknown, infer V> ? V : never

export { Mutation, MutationFunction, useMutation } from "./mutation"
export { Query, useQuery } from "./query"
