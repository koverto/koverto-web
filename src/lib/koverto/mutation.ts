import { Mutation } from "react-apollo-typed-hooks"
import { Authentication, LoginResponse, User, UserInput } from "./schema"

export class CreateUser extends Mutation<
  { createUser: LoginResponse },
  { input: Authentication }
> {}

export class Login extends Mutation<
  { login: LoginResponse },
  { input: Authentication }
> {}

export class UpdateUser extends Mutation<
  { updateUser: User },
  { input: UserInput }
> {}
