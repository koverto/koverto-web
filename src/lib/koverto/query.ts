import { Query } from "react-apollo-typed-hooks"
import { User } from "./schema"

export class GetUser extends Query<{ getUser: User }, {}> {}
