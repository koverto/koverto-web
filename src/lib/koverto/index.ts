import * as Mutation from "./mutation"
import * as Query from "./query"

export * from "./schema"
export { Mutation, Query }
