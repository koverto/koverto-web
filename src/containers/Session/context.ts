import { Dispatch, SetStateAction, createContext } from "react"
import { ServerError } from "apollo-link-http-common"
import { ErrorResponse } from "apollo-link-error"
import { SessionProviderProps } from "./provider"

export interface SessionProps extends SessionProviderProps {
  isLoggedIn?: () => boolean
  setToken?: Dispatch<SetStateAction<string>>
}

export class Session implements SessionProps {
  private readonly _basePath: string
  private readonly _loginPath: string
  public readonly isLoggedIn: () => boolean
  public readonly setToken: Dispatch<SetStateAction<string>>

  constructor({
    basePath = "/",
    loginPath = "/login",
    isLoggedIn = (): boolean => false,
    setToken = (_): void => null,
  }: SessionProps = {}) {
    this._basePath = basePath
    this._loginPath = loginPath
    this.isLoggedIn = isLoggedIn
    this.setToken = setToken
  }

  get basePath(): string {
    return this._basePath
  }

  get loginPath(): string {
    return this._loginPath
  }

  public handleError = ({ networkError = null }: ErrorResponse): void => {
    if ((networkError as ServerError)?.statusCode === 401) {
      this.setToken(null)
    }
  }
}

export const SessionContext = createContext<Session>(new Session())
