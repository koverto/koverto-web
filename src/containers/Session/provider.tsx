import { ApolloProvider } from "@apollo/react-hooks"
import * as React from "react"
import { Session, SessionContext } from "./context"
import { useStateFromLocalStorage } from "../../lib/local-storage-state"
import { ApolloLink } from "apollo-link"
import { AuthLink } from "../../lib/apollo-auth-link"
import { ErrorLink } from "apollo-link-error"
import { InMemoryCache } from "apollo-cache-inmemory"
import { HttpLink } from "apollo-link-http"
import { ApolloClient } from "apollo-client"

const cache = new InMemoryCache()
const httpLink = new HttpLink({ uri: GRAPHQL_ENDPOINT })

export interface SessionProviderProps {
  basePath?: string
  loginPath?: string
}

export const SessionProvider = ({
  basePath = "/",
  loginPath = "/login",
  children,
}: React.PropsWithChildren<SessionProviderProps>): JSX.Element => {
  const [token, setToken] = useStateFromLocalStorage<string>("token")

  const getToken = React.useCallback(() => token, [token])
  const isLoggedIn = React.useCallback(() => !!token, [token])
  const session = new Session({ basePath, loginPath, isLoggedIn, setToken })

  const link = ApolloLink.from([
    new AuthLink(getToken),
    new ErrorLink(session.handleError),
    httpLink,
  ])

  const client = new ApolloClient({ cache, link })

  return (
    <SessionContext.Provider value={session}>
      <ApolloProvider client={client}>{children}</ApolloProvider>
    </SessionContext.Provider>
  )
}
