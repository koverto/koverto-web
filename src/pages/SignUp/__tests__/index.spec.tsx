import { MockedProvider } from "@apollo/react-testing"
import {
  render as _render,
  RenderResult,
  waitFor,
} from "@testing-library/react"
import * as React from "react"
import { MemoryRouter } from "react-router-dom"
import { onSubmit as _onSubmit, QUERY, SignUp } from ".."
import { Session, SessionContext } from "../../../containers/Session/context"
import { CredentialType } from "../../../lib/koverto"

describe("onSubmit", () => {
  const createUser = jest.fn()
  const onSubmit = _onSubmit(createUser)

  it("builds the mutation parameters from the form data", () => {
    onSubmit({ email: "test@test", name: "test", password: "P@55w0rd!" })

    expect(createUser).toBeCalledWith({
      variables: {
        input: {
          user: { email: "test@test", name: "test" },
          credential: {
            credentialType: CredentialType.PASSWORD,
            credential: "P@55w0rd!",
          },
        },
      },
    })
  })
})

const setToken = jest.fn()
const render = (result = {}): RenderResult =>
  _render(
    <SessionContext.Provider value={new Session({ setToken })}>
      <MockedProvider
        addTypename={false}
        mocks={[
          {
            request: {
              query: QUERY,
              variables: {
                input: {
                  user: { email: "test@test", name: "test" },
                  credential: {
                    credentialType: CredentialType.PASSWORD,
                    credential: "P@55w0rd!",
                  },
                },
              },
            },
            result,
          },
        ]}
      >
        <MemoryRouter>
          <SignUp />
        </MemoryRouter>
      </MockedProvider>
    </SessionContext.Provider>
  )

describe("SignUp", () => {
  beforeEach(() => {
    localStorage.clear()
  })

  afterEach(() => {
    localStorage.clear()
  })

  it("matches the snapshot", () => {
    const { container } = render()
    expect(container).toMatchSnapshot()
  })

  it("does not set the token if there is no mutation result data", () => {
    render()
    expect(localStorage.setItem).not.toBeCalled()
  })

  it("does not set the token if there is mutation result data containing an error", async () => {
    const { findByTestId } = render({
      errors: [
        {
          message: "could not create user",
          path: ["createUser"],
        },
      ],
    })

    const email = await findByTestId("email")
    const name = await findByTestId("name")
    const password = await findByTestId("password")
    const button = await findByTestId("submit")

    email.setAttribute("value", "test@test")
    name.setAttribute("value", "test")
    password.setAttribute("value", "P@55w0rd!")
    button.click()

    await findByTestId("signup-error")
    expect(localStorage.setItem).not.toBeCalled()
  })

  it("sets the token if one is included with the mutation result data", async () => {
    const { findByTestId } = render({
      data: {
        createUser: {
          token: "token",
          user: {
            id: "7527500c-1238-496d-aba8-1d21b6325f83",
            name: "test",
            email: "test@test",
          },
        },
      },
    })

    const email = await findByTestId("email")
    const name = await findByTestId("name")
    const password = await findByTestId("password")
    const button = await findByTestId("submit")

    email.setAttribute("value", "test@test")
    name.setAttribute("value", "test")
    password.setAttribute("value", "P@55w0rd!")
    button.click()

    await waitFor(() => expect(setToken).toBeCalledWith("token"))
  })
})
