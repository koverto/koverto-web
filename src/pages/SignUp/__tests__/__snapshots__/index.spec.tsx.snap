// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp matches the snapshot 1`] = `
<div>
  <div>
    <h1>
      Sign Up
    </h1>
    <form>
      <div>
        <input
          data-testid="name"
          name="name"
          type="text"
        />
      </div>
      <div>
        <input
          data-testid="email"
          name="email"
          type="email"
        />
      </div>
      <div>
        <input
          data-testid="password"
          name="password"
          type="password"
        />
      </div>
      <button
        data-testid="submit"
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</div>
`;
